File: scripts/responsive_audit.py
#!/usr/bin/env python3
"""Responsive layout audit for all main prototype pages."""
from __future__ import annotations

import json
import sys
from playwright.sync_api import sync_playwright

BASE = "http://127.0.0.1:8090"
PAGES = [
    ("/index.html", "Home"),
    ("/search.html", "Search"),
    ("/search.html?q=tebow", "Search (Tebow)"),
    ("/happenings.html", "Happenings outline"),
    ("/happenings-listing.html", "Listing"),
    ("/happenings-transactional.html?event=gameday-kickoff", "Transactional"),
    ("/happenings-rsvp.html?event=football-fest", "RSVP (Fest)"),
    ("/ccfb-logo-options.html", "CCFB logos"),
    ("/happenings-hero-options.html", "Hero options"),
    ("/happenings-form-embed.html", "Form embed"),
]
WIDTHS = [390, 768, 901, 1100, 1280, 1440]

AUDIT_JS = """
() => {
  const doc = document.documentElement;
  const overflow = doc.scrollWidth - doc.clientWidth;
  const inner = document.querySelector('.site-footer__inner');
  const know = document.querySelector('.site-footer__col--know');
  const rail = document.querySelector('.site-rail');
  const topbar = document.querySelector('.mobile-topbar');
  const banner = document.querySelector('.hp-banner');
  let worst = null;
  for (const el of document.querySelectorAll('body *')) {
    const r = el.getBoundingClientRect();
    if (r.width < 2 || r.height < 2) continue;
    const o = r.right - window.innerWidth;
    if (o > 2) {
      const cls = el.className && typeof el.className === 'string'
        ? el.className.split(/\\s+/).slice(0, 2).join('.')
        : el.tagName.toLowerCase();
      if (!worst || o > worst.o) worst = { cls, o: Math.round(o) };
    }
  }
  const rs = rail ? getComputedStyle(rail) : null;
  const railHidden = rs && rs.position === 'fixed' && rs.transform !== 'none'
    && !rs.transform.includes('matrix(1, 0, 0, 1, 0, 0)');
  return {
    overflow: Math.round(overflow),
    worst,
    topbarOn: topbar ? getComputedStyle(topbar).display !== 'none' : null,
    railHidden,
    bannerOverflow: banner ? banner.scrollWidth > banner.clientWidth + 1 : false,
    footerGrid: inner ? getComputedStyle(inner).gridTemplateColumns : null,
    knowFlex: know ? getComputedStyle(know).flexDirection : null,
    knowSpan: know ? getComputedStyle(know).gridColumn : null,
  };
}
"""


def classify(page: str, width: int, a: dict) -> list[dict]:
    issues = []
    if a["overflow"] > 1:
        detail = f"{a['overflow']}px overflow"
        if a.get("worst"):
            detail += f" — {a['worst']['cls']} (+{a['worst']['o']}px)"
        issues.append({"severity": "P1", "type": "horizontal-overflow", "detail": detail})
    if a.get("bannerOverflow"):
        issues.append({"severity": "P2", "type": "banner-overflow", "detail": "hp-banner wider than viewport"})
    if width <= 900 and a.get("topbarOn") is False:
        issues.append({"severity": "P1", "type": "missing-mobile-topbar", "detail": "No mobile topbar at ≤900px"})
    if width > 900 and a.get("railHidden") is True:
        issues.append({"severity": "P2", "type": "rail-hidden", "detail": "Rail off-screen at desktop width"})
    if (
        width == 1100
        and a.get("footerGrid")
        and "300px" in a["footerGrid"]
        and a.get("knowSpan") == "auto"
        and a.get("knowFlex") == "row"
    ):
        issues.append({
            "severity": "P1",
            "type": "footer-overlap-risk",
            "detail": f"3-col + row know at 1100px: {a['footerGrid']}",
        })
    return issues


def main() -> int:
    rows = []
    issues = []

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        for path, name in PAGES:
            for w in WIDTHS:
                page.set_viewport_size({"width": w, "height": 900})
                page.goto(f"{BASE}{path}", wait_until="networkidle", timeout=30000)
                page.wait_for_timeout(300)
                page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
                page.evaluate("window.scrollTo(0, 0)")
                a = page.evaluate(AUDIT_JS)
                row = {"page": name, "path": path, "width": w, **a}
                rows.append(row)
                for issue in classify(name, w, a):
                    issues.append({"page": name, "path": path, "width": w, **issue})
        browser.close()

    out = {"rows": rows, "issues": issues}
    print(json.dumps(out, indent=2))
    print(f"\nAudit: {len(issues)} issue(s)", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/responsive_audit.mjs
#!/usr/bin/env node
/**
 * Responsive audit via CDP — run with Cursor browser tab open, or:
 *   node scripts/responsive_audit.mjs
 * Uses fetch to localhost; for layout checks requires puppeteer.
 */
import puppeteer from 'puppeteer';

const BASE = process.env.AUDIT_BASE || 'http://127.0.0.1:8090';
const pages = [
  { path: '/index.html', name: 'Home' },
  { path: '/search.html', name: 'Search' },
  { path: '/search.html?q=tebow', name: 'Search (Tebow)' },
  { path: '/happenings.html', name: 'Happenings outline' },
  { path: '/happenings-listing.html', name: 'Listing' },
  { path: '/happenings-transactional.html?event=gameday-kickoff', name: 'Transactional' },
  { path: '/happenings-rsvp.html?event=football-fest', name: 'RSVP (Fest)' },
  { path: '/ccfb-logo-options.html', name: 'CCFB logos' },
  { path: '/happenings-hero-options.html', name: 'Hero options' },
  { path: '/happenings-form-embed.html', name: 'Form embed' },
];
const widths = [390, 768, 901, 1100, 1280, 1440];

function auditPage() {
  const doc = document.documentElement;
  const overflow = doc.scrollWidth - doc.clientWidth;
  const footer = document.querySelector('.site-footer');
  const inner = document.querySelector('.site-footer__inner');
  const know = document.querySelector('.site-footer__col--know');
  const rail = document.querySelector('.site-rail');
  const topbar = document.querySelector('.mobile-topbar');
  const banner = document.querySelector('.hp-banner');

  const offenders = [];
  for (const el of document.querySelectorAll('body *')) {
    const r = el.getBoundingClientRect();
    if (r.width < 2 || r.height < 2) continue;
    if (r.right > window.innerWidth + 2) {
      const cls =
        el.className && typeof el.className === 'string'
          ? el.className.split(/\s+/).slice(0, 2).join('.')
          : el.tagName.toLowerCase();
      offenders.push({
        cls,
        overflow: Math.round(r.right - window.innerWidth),
      });
    }
  }
  offenders.sort((a, b) => b.overflow - a.overflow);
  const top = [];
  const seen = new Set();
  for (const o of offenders) {
    const k = `${o.cls}:${o.overflow}`;
    if (seen.has(k)) continue;
    seen.add(k);
    top.push(o);
    if (top.length >= 3) break;
  }

  const railStyle = rail ? getComputedStyle(rail) : null;
  const railOffscreen =
    railStyle &&
    railStyle.position === 'fixed' &&
    railStyle.transform !== 'none' &&
    !railStyle.transform.includes('matrix(1, 0, 0, 1, 0, 0)');

  return {
    overflow,
    hasOverflow: overflow > 1,
    topbarOn: topbar ? getComputedStyle(topbar).display !== 'none' : null,
    railHidden: rail ? railOffscreen : null,
    bannerOverflow: banner ? banner.scrollWidth > banner.clientWidth + 1 : false,
    footerGrid: inner ? getComputedStyle(inner).gridTemplateColumns : null,
    knowFlex: know ? getComputedStyle(know).flexDirection : null,
    knowSpan: know ? getComputedStyle(know).gridColumn : null,
    offenders: top,
  };
}

const browser = await puppeteer.launch({ headless: true, args: ['--no-sandbox'] });
const pg = await browser.newPage();
const rows = [];
const issues = [];

for (const page of pages) {
  for (const w of widths) {
    await pg.setViewport({ width: w, height: 900 });
    await pg.goto(`${BASE}${page.path}`, { waitUntil: 'networkidle2', timeout: 30000 });
    await pg.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
    await pg.evaluate(() => window.scrollTo(0, 0));
    const a = await pg.evaluate(auditPage);
    rows.push({ page: page.name, width: w, ...a });

    if (a.hasOverflow) {
      issues.push({
        severity: 'P1',
        page: page.name,
        width: w,
        type: 'horizontal-overflow',
        detail: `${a.overflow}px — ${a.offenders.map((o) => o.cls).join(', ')}`,
      });
    }
    if (a.bannerOverflow) {
      issues.push({
        severity: 'P2',
        page: page.name,
        width: w,
        type: 'banner-overflow',
        detail: 'hp-banner wider than viewport',
      });
    }
    if (w <= 900 && a.topbarOn === false) {
      issues.push({
        severity: 'P1',
        page: page.name,
        width: w,
        type: 'missing-mobile-topbar',
        detail: 'Expected mobile topbar at ≤900px',
      });
    }
    if (w > 900 && a.railHidden === true) {
      issues.push({
        severity: 'P2',
        page: page.name,
        width: w,
        type: 'rail-hidden',
        detail: 'Rail off-screen at desktop width',
      });
    }
    if (w === 1100 && a.footerGrid && a.footerGrid.includes('300px') && a.knowSpan === 'auto') {
      issues.push({
        severity: 'P1',
        page: page.name,
        width: w,
        type: 'footer-overlap-risk',
        detail: `3-col footer at 1100px: ${a.footerGrid}`,
      });
    }
  }
}

await browser.close();

const out = { rows, issues };
console.log(JSON.stringify(out, null, 2));
console.error(`\nAudit complete: ${issues.length} issue(s) across ${pages.length} pages × ${widths.length} widths`);
process.exit(issues.length > 0 ? 0 : 0);
